server/auth/range_perm_cache_test.go

```diff
@@ -100,3 +130,85 @@ func TestKeyPermission(t *testing.T) {
 		}
 	}
 }
+
+func TestRangeCheck(t *testing.T) {
+	tests := []struct {
+		key      []byte
+		rangeEnd []byte
+		want     bool
+	}{
+		{
+			// valid single key
+			[]byte("a"),
+			[]byte(""),
+			true,
+		},
+		{
+			// valid single key
+			[]byte("a"),
+			nil,
+			true,
+		},
+		{
+			// valid key range, key < rangeEnd
+			[]byte("a"),
+			[]byte("b"),
+			true,
+		},
+		{
+			// invalid empty key range, key == rangeEnd
+			[]byte("a"),
+			[]byte("a"),
+			false,
+		},
+		{
+			// invalid empty key range, key > rangeEnd
+			[]byte("b"),
+			[]byte("a"),
+			false,
+		},
+		{
+			// invalid key, key must not be ""
+			[]byte(""),
+			[]byte("a"),
+			false,
+		},
+		{
+			// invalid key range, key must not be ""
+			[]byte(""),
+			[]byte(""),
+			false,
+		},
+		{
+			// invalid key range, key must not be ""
+			[]byte(""),
+			[]byte("\x00"),
+			false,
+		},
+		{
+			// valid single key (not useful in practice)
+			[]byte("\x00"),
+			[]byte(""),
+			true,
+		},
+		{
+			// valid key range, larger or equals to "a"
+			[]byte("a"),
+			[]byte("\x00"),
+			true,
+		},
+		{
+			// valid key range, which includes all keys
+			[]byte("\x00"),
+			[]byte("\x00"),
+			true,
+		},
+	}
+
+	for i, tt := range tests {
```

Comment:
The comment https://github.com/etcd-io/etcd/pull/15294#discussion_r1111372666 applies here as well.

Reply:
ah sorry, let me update here too